File: ruby/sinatra_graphql/lib/api/types/user_type.rb
UserType = GraphQL::ObjectType.define do
  name "User"
  description "A user"

  field :id, !types.ID
  field :first_name, !types.String
  field :last_name, !types.String
  field :username, !types.String
  field :email, !types.String
  field :posts, -> { types[!PostType] }
  field :comments, -> { types[!CommentType] }
end

File: javascript/koa_graphql/api/types/userType.js
import PostType from './postType';
import CommentType from './commentType';

import {
  GraphQLBoolean,
  GraphQLID,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
} from 'graphql';

const UserType = new GraphQLObjectType({
  name: 'UserType',
  fields: () => ({
    id: { type: new GraphQLNonNull(GraphQLID) },
    first_name: {
      type: GraphQLString,
      resolve: (obj) => obj.first_name,
    },
    last_name: {
      type: GraphQLString,
      resolve: (obj) => obj.last_name,
    },
    username: {
      type: GraphQLString,
      resolve: (obj) => obj.username,
    },
    email: {
      type: GraphQLString,
      resolve: (obj) => obj.email,
    },
    posts: {
      type: new GraphQLList(PostType),
      resolve: (obj) => obj.posts
    },
    comments: {
      type: new GraphQLList(CommentType),
      resolve: (obj) => obj.comments
    },
  }),
});

export default UserType;
